### PR TITLE
Feature BL-15043 linting and static analysis

### PR DESCRIPTION
**Pull request by waringr**
## Description

Adds phpcs, and phpstan analysis to the pre-commit hook and fixes all errors.

Related issue: https://dvsa.atlassian.net/browse/BL-15043

## Before submitting (or marking as "ready for review")

+ [ ] Does the pull request title follow the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification?
+ [ ] Have you performed a self-review of the code
+ [ ] Have you have added tests that prove the fix or feature is effective and working
+ [ ] Did you make sure to update any documentation relating to this change?

### CODE REVIEW

File: src/CpmsCommon/AbstractInputFilterFactory.php [hunk 1]
```diff
@@ -64,16 +64,18 @@ public function getClassName($name, ContainerInterface $serviceLocator)
...
-            $routeMatch = $serviceLocator->get('Application')->getMvcEvent()->getRouteMatch();
+            /** @var Application */
+            $application = $serviceLocator->get('Application');
+            /** @var ?RouteMatch $routeMatch */
+            $routeMatch = $application->getMvcEvent()->getRouteMatch();
 
             if (empty($routeMatch)) {
                 $version = $defaultVersion;
             } else {
-                $version = (int)$routeMatch->getParam('version');
```
Comment (aaronlong) [inline src/CpmsCommon/AbstractInputFilterFactory.php:76]: Do we know what format the version param looks like currently?  My question essentially boils down to was the casting making this work in the past?

File: phpstan.neon [hunk 2]
```diff
@@ -1,7 +1,9 @@
...
-    checkMissingIterableValueType: false
-    checkGenericClassInNonGenericObjectType: false
+    ignoreErrors:
+        - identifier: missingType.iterableValue
+        - identifier: missingType.generics
     level: max
     paths:
         - src
         - test
+    treatPhpDocTypesAsCertain: false
```
Comment (UndermountainK) [inline phpstan.neon]: Any particular reason for this parameter? In other repos we did not have to add this

File: src/CpmsCommon/Controller/Plugin/SendPayload.php [hunk 3]
```diff
@@ -18,29 +20,31 @@
...
-        /** @var \Laminas\Http\PhpEnvironment\Response $response */
-        /** @var RouteMatch $routeMatch */
-        /** @var Application $application */
         $controller  = $this->getController();
         $serviceLocator = $controller->getServiceLocator();
 
-        $payLoad     = (array)$payLoad;
-        $config      = $serviceLocator->get('config');
+        $payLoad = (array)$payLoad;
+        /** @var array */
```
Comment (UndermountainK) [inline src/CpmsCommon/Controller/Plugin/SendPayload.php]: why sometimes we provide the variable name and sometimes not in doc strings?

**Comment by UndermountainK** [comment]
Will we be testing this by pulling this package to:
[dvsa/mot-cpms-shared-reports](https://github.com/dvsa/mot-cpms-shared-reports)
[dvsa/mot-cpms-payment-test](https://github.com/dvsa/mot-cpms-payment-test)
[dvsa/mot-cpms-forms](https://github.com/dvsa/mot-cpms-forms)

and after that to MTS to make sure everything is working?

File: src/CpmsCommon/Queue/Adapter/Synchronous/SynchronousQueueAdapterFactory.php [hunk 4]
```diff
@@ -2,32 +2,29 @@
...
-use Laminas\ServiceManager\Factory\FactoryInterface;
+use Laminas\ServiceManager\ServiceLocatorInterface;
 
 /**
  * Class SynchronousQueueAdapterFactory
  *
  * @package Queue\Adapter\Immediate
  */
-class SynchronousQueueAdapterFactory implements FactoryInterface
+class SynchronousQueueAdapterFactory
```
Comment (UndermountainK) [inline src/CpmsCommon/Queue/Adapter/Synchronous/SynchronousQueueAdapterFactory.php:15]: Why the change here? I think every Factory should follow this pattern  with Psr/ContainerInterface

File: src/CpmsCommon/Service/ApiContentTypeFactory.php [hunk 5]
```diff
@@ -1,35 +1,30 @@
...
-use Laminas\ServiceManager\Factory\FactoryInterface;
-use Laminas\ServiceManager\ServiceLocatorInterface;
 
 /**
  * Class ApiVersion
  *
  * @package CpmsCommon\Service
  */
-class ApiContentTypeFactory implements FactoryInterface
+class ApiContentTypeFactory
```
Comment (UndermountainK) [inline src/CpmsCommon/Service/ApiContentTypeFactory.php:12]: same as above

File: src/CpmsCommon/Service/Config/AuthServiceOptions.php [hunk 6]
```diff
@@ -21,7 +22,7 @@ class AuthServiceOptions extends AbstractOptions
     /**
      * The scope required for the API endpoint
      *
-     * @var string
+     * @var string|bool
```
Comment (UndermountainK) [inline src/CpmsCommon/Service/Config/AuthServiceOptions.php]:
```suggestion
     * @var string|bool|null
```

File: src/CpmsCommon/Service/ProfilingInitializer.php [hunk 7]
```diff
@@ -7,7 +8,7 @@
 
 namespace CpmsCommon\Service;
 
-use Interop\Container\ContainerInterface;
+use Psr\Container\ContainerInterface;
```
Comment (UndermountainK) [inline src/CpmsCommon/Service/ProfilingInitializer.php:11]: If we are changing to a newer Psr ContainerInterface can we do it for every factory? So that old Interop wont be used anymore

File: test/CpmsCommonTest/Utility/LoggerAwareTraitTest.php [hunk 8]
```diff
@@ -3,13 +3,14 @@
 namespace CpmsCommonTest\Utility;
 
 use CpmsCommonTest\Bootstrap;
+use Laminas\ServiceManager\ServiceManager;
 
 class LoggerAwareTraitTest extends \PHPUnit\Framework\TestCase
 {
-
+    /** @phpstan-ignore missingType.property */
```
Comment (UndermountainK) [inline test/CpmsCommonTest/Utility/LoggerAwareTraitTest.php]: Should we remove it and add LoggerInterface type? unless i am wrong and we cannot find out about the exact type

File: src/CpmsCommon/AbstractInputFilterFactory.php (hunk 1 above)
Comment (waringr) [reply src/CpmsCommon/AbstractInputFilterFactory.php:76]: The phpstan issue is that you can't cast `mixed` to int. It appears that you also can't `intval()` mixed to int. I've added a check for '0' on the following line to cover that off.

File: phpstan.neon (hunk 2 above)
Comment (waringr) [reply phpstan.neon]: It was to cover off a couple of errors. I've reduced that number to one, but I think we're a bit stuck because of some dodgy typing in Laminas. See `src/CpmsCommon/View/JsonExceptionStrategy.php` line 112. If you click through to the MVCEvent, then it only returns `ApplicationInterface` for `getApplication`, rather than `ApplicationInterface|null`.

Comment (waringr) [reply phpstan.neon]: I've removed the config and will ignore the above for now, unless anyone has any suggestions?

File: src/CpmsCommon/Controller/Plugin/SendPayload.php (hunk 3 above)
Comment (waringr) [reply src/CpmsCommon/Controller/Plugin/SendPayload.php]: I've gone through and standardised it to a single line with the variable name. Class vars now have the type as part of the declaration, not in a comment.

File: src/CpmsCommon/Queue/Adapter/Synchronous/SynchronousQueueAdapterFactory.php (hunk 4 above)
Comment (waringr) [reply src/CpmsCommon/Queue/Adapter/Synchronous/SynchronousQueueAdapterFactory.php:15]: Because the `setServiceContainer` call on the body of the factory needs a ServiceLocatorInterface (vendor/laminas/laminas-servicemanager/src/ServiceLocatorInterface.php), which is a child of the `ContainerInterface`. So the signatures don't match.

File: src/CpmsCommon/Service/ApiContentTypeFactory.php (hunk 5 above)
Comment (waringr) [reply src/CpmsCommon/Service/ApiContentTypeFactory.php:12]: Similar to above, but here it's because this factory is returning a string, not an object, so it doesn't match the interface return value.

File: src/CpmsCommon/Service/Config/AuthServiceOptions.php (hunk 6 above)
Comment (waringr) [reply src/CpmsCommon/Service/Config/AuthServiceOptions.php]: Done :)

File: src/CpmsCommon/Service/ProfilingInitializer.php (hunk 7 above)
Comment (waringr) [reply src/CpmsCommon/Service/ProfilingInitializer.php:11]: Done

File: test/CpmsCommonTest/Utility/LoggerAwareTraitTest.php (hunk 8 above)
Comment (waringr) [reply test/CpmsCommonTest/Utility/LoggerAwareTraitTest.php]: It just freaks out because (I assume) the trait isn't designed to be used on it's own and the`getMockForTrait` call lower down just returns a MockObject with no indication that the trait is attached to it.

Comment (waringr) [reply test/CpmsCommonTest/Utility/LoggerAwareTraitTest.php]: I've added a new interface and a new mock for the trait, which has fixed it.

File: src/CpmsCommon/AbstractService.php [hunk 9]
```diff
@@ -1,57 +1,53 @@
...
     }
 
     /**
-     * @param ContainerInterface $serviceLocator
+     * @param ServiceLocatorInterface $serviceLocator
      * @return AbstractService
      */
-    public function setServiceLocator(ContainerInterface $serviceLocator)
+    public function setServiceLocator($serviceLocator)
     {
```
Comment (PapaZouk) [inline src/CpmsCommon/AbstractService.php:51]: I believed that in the latest PHP version, defining return types at the declaration level is recommended rather than in doc blocks. Was there a reason for this change?

File: src/CpmsCommon/Service/Config/AuthServiceOptions.php
Comment (PapaZouk) [inline src/CpmsCommon/Service/Config/AuthServiceOptions.php:1]: I believe we need to be consistent with typing since we add types at the definition level in one file and in the doc block in another. Should we adopt a single approach across all files?

File: src/CpmsCommon/AbstractInputFilterFactory.php [hunk 10]
```diff
@@ -63,18 +63,20 @@ public function getClassName($name, ContainerInterface $serviceLocator)
             $namespaceParts[$i] = ucfirst($namespaceParts[$i]);
         }
 
-        if (count($namespaceParts) > 0 and in_array($namespaceParts[0], $this->versionedNamespaces)) {
-
-            $config         = $serviceLocator->get('config');
+        if (in_array($namespaceParts[0], $this->versionedNamespaces)) {
```
Comment (UndermountainK) [inline src/CpmsCommon/AbstractInputFilterFactory.php:66]: I believe the behaviour here is now changed since we dont check if var is more than 0.

Comment (waringr) [reply src/CpmsCommon/AbstractInputFilterFactory.php:66]: It was complaining that the `count($namespaceParts)` would never be zero, I checked it in the PHP REPL and it will always have a length of at least 1, because the explode will give an empty string, even if it's passed an empty string. So that `count` test is never false. 

Based on that, I think we're ok.

File: src/CpmsCommon/AbstractService.php (hunk 9 above)
Comment (waringr) [reply src/CpmsCommon/AbstractService.php:51]: I think it was for consistency within the file and that PHPCS/Phpstan didn't complain about it. I'll go for declaration level rather than docblock

File: src/CpmsCommon/Service/Config/AuthServiceOptions.php
Comment (waringr) [reply src/CpmsCommon/Service/Config/AuthServiceOptions.php:1]: As above, all declared in definition now.

**Review by UndermountainK — APPROVED** [review]

**Review by PapaZouk — APPROVED** [review]
